src/EnergyPlus/Construction.cc

```diff
@@ -2063,4 +2062,128 @@ void ConstructionProps::setArraysBasedOnMaxSolidWinLayers(EnergyPlusData &state)
+
+        // get the cached data if it exists. if it doesn't nlohmann::json throws out_of_range
+        nlohmann::json thisConstrData = state.dataCache->unorderedCTFObjects.at(key);
+
+        // nlohmann::json does some weird things when directly accessing strings
+        // explicitly setting these as string objects here so we can do a proper comparison
+        std::string a = thisConstrData.at("full_data_key");
+        std::string b = this->getCacheKeyString(state);
+
+        // final confirmation that the object we found matches exactly with the current construction
+        if (a != b) {
+            // show warning message here?
+            this->CTFLoadedFromCache = false;
+            return;
+        }
+
+        // we can load the data if if we've made it this far
+
+        // CTF arrays
+        Cache::jsonToArray(state, this->CTFCross, thisConstrData, "ctf_cross");
+        Cache::jsonToArray1(state, this->CTFFlux, thisConstrData, "ctf_flux");
+        Cache::jsonToArray(state, this->CTFInside, thisConstrData, "ctf_inside");
+        Cache::jsonToArray(state, this->CTFOutside, thisConstrData, "ctf_outside");
+
+        // other necessary data
+        Cache::jsonToData(state, this->NumHistories, thisConstrData, "num_histories");
+        Cache::jsonToData(state, this->NumCTFTerms, thisConstrData, "num_ctf_terms");
+        Cache::jsonToData(state, this->CTFTimeStep, thisConstrData, "ctf_timestep");
+        Cache::jsonToData(state, this->UValue, thisConstrData, "u_value");
+
+        this->CTFLoadedFromCache = true;
+    } catch (const nlohmann::json::out_of_range &e) {
+        // should already be defaulted to false, but this makes sure
+        this->CTFLoadedFromCache = false;
+    }
+}
+
+std::string ConstructionProps::getCacheKey(EnergyPlusData &state)
+{
+    unsigned long long key = 0;
+
+    // data from material layers
+    // adding the layer num to get a unique key for objects with reversed layers
+    for (int Layer = 1; Layer <= this->TotLayers; ++Layer) {
+        int CurrentLayer = this->LayerPoint(Layer);
+        auto &mat = state.dataMaterial->Material(CurrentLayer);
+
+#ifdef USE_REINTERPRET_CAST
+        key ^= Cache::prepFloatForCacheKey(mat.Thickness, 15) + Layer;
+        key ^= Cache::prepFloatForCacheKey(mat.Conductivity, 15) + Layer;
+        key ^= Cache::prepFloatForCacheKey(mat.Density, 15) + Layer;
+        key ^= Cache::prepFloatForCacheKey(mat.SpecHeat, 15) + Layer;
+        key ^= Cache::prepFloatForCacheKey(mat.Resistance, 15) + Layer;
+#endif
+#ifdef USE_STATIC_CAST_AND_FP_MULT
+        key ^= Cache::prepFloatForCacheKey(mat.Thickness, 1E5) + Layer;
+        key ^= Cache::prepFloatForCacheKey(mat.Conductivity, 1E5) + Layer;
+        key ^= Cache::prepFloatForCacheKey(mat.Density, 1E5) + Layer;
+        key ^= Cache::prepFloatForCacheKey(mat.SpecHeat, 1E5) + Layer;
+        key ^= Cache::prepFloatForCacheKey(mat.Resistance, 1E5) + Layer;
+#endif
+#ifdef USE_C_CAST_AND_FP_MULT
```

Comment:
This block looks identical to previous one. Can combine using `#if defined(USE_STATIC_CAST_AND_FP_MULT) || defined(USE_C_CAST_AND_FP_MULT)`. Also, I am not entirely sure that there is any difference between C cast and `static_cast` for scalar data types. In this specific case, I believe those two are the same option, so this may not be necessary anyway.

Reply:
Thanks for pointing that out. I think these ```#ifdef```s are going away before this PR merges. I just wanted to give anyone interested an opportunity to comment on the different approaches.